### PR TITLE
documentation: Document test suite regeneration

### LINKED ISSUE #717: Document test regeneration.
We need to document the test re-generation procedure so that it is possible for new contributors to do.

Future goal: automate it.

### PR DESCRIPTION
**Pull request by rpalo**
Wrote some documentation explaining how to regenerate test data and other data for exercises.  I also included some changes to explain when to use `-u` to update the VERSION number as well.  Lastly, I added a section to the *Contributing* section that gives some ideas for getting started to newcomers, including the newly revamped *Regenerating Test Suites* section.

Let me know what you think.  I'm happy to do some revising to get the wording just right.

## Why?
Solves #717, as we discussed in that issue (see also some of the discussion in #757 PR.

## How Has This Been Tested?
No tests.  All I did was change the README.  However, the methodology that I wrote down is sound, because I just did it that way a couple of days ago 😄 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or enhancement that would cause existing functionality to change)
- [x] Gardening (code and/or documentation cleanup)

## References and Closures

References #757 
Closes #717 

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change relies on a pending issue/pull request
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Insti** [comment]
Awesome, thanks! :heart: I'll have a look through this later when I get a chance.

File: README.md [hunk 1]
```diff
@@ -104,12 +102,29 @@ tree -L 1 ~/code/exercism
...
-    bin/generate --update <slug> 
+If it's your first time cloning/contributing to the repository, you'll need to install any dependencies via `bundle`:
 
-Leaving out the --update option will cause the BookKeeping version number to remain the same.
-This can be useful when testing generators.
+```bash
+bundle install
+```
+
+The generator also depends on the presence of Exercism's `problem-specifications` repository (see the file tree in the section above).  Make sure you've got an *up-to-date* version of the specifications in a `problem-specifications` folder that's in a parallel directory to your local copy of the `ruby` repository.
```
Comment (Insti) [inline README.md]: How do I make sure my repository is up to date?

File: README.md [hunk 2]
```diff
@@ -104,12 +102,29 @@ tree -L 1 ~/code/exercism
...
+
+To check which problems have possibly been updated, run:
+
+```bash
+bin/generate -a
```
Comment (Insti) [inline README.md]: Use the full flag: `--all`

File: README.md [hunk 3]
```diff
@@ -104,12 +102,29 @@ tree -L 1 ~/code/exercism
...
+```
+
+This will autogenerate all of the exercises that can be.  Use your editor, diff tool, or `git status` to find out which test files have changed.  Some exercises will update because someone updated the description or other exercise metadata.  Others will change because the actual test suite has changed.  If you find that an exercise's test suite (i.e. the actual tests, not just the line at the test data version number at the top of the tests) has changed, be sure to use `generate` to auto-increment the [BookKeeping::VERSION](#bookkeeping-version) number by running:
```
Comment (Insti) [inline README.md]: > all of the exercises that can be.

all of the test cases for which generators exist.

Comment (Insti) [reply README.md]: > Use your editor, diff tool, or `git status` to find out which test files have changed.

Too many options here. 
`git diff` is a useful command to use to see what has changed.

Comment (Insti) [reply README.md]: > to auto-increment

to update

File: README.md [hunk 4]
```diff
@@ -104,12 +102,29 @@ tree -L 1 ~/code/exercism
...
+
+```bash
+bin/generate -u <exercise-slug>
```
Comment (Insti) [inline README.md]: Use the full flag: `--update`

File: README.md [hunk 5]
```diff
@@ -104,12 +102,29 @@ tree -L 1 ~/code/exercism
...
+```
+
+Once everything has been regenerated and updated, you're almost ready to submit your changes via pull request.  Please follow the guidelines in the [Pull Requests](#pull-requests) section, being sure to follow the pattern of `<slug>: Regenerate Tests`, where slug is the slug of the one exercise that your pull request is regenerating.
```
Comment (Insti) [inline README.md]: Add something about: Only update one problem per pull request.

File: README.md [hunk 6]
```diff
@@ -260,10 +275,8 @@ are constructed using shared metadata, which lives in the [problem-specification
 
 ## Contributing Guide
 
-For an in-depth discussion of how exercism language tracks and exercises work,
-please see the
-[contributing guide](https://github.com/exercism/x-api/blob/master/CONTRIBUTING.md#the-exercise-data)
-
+For an in-depth discussion of how exercism language tracks and exercises work, please see the
+[contributing guide](https://github.com/exercism/x-api/blob/master/CONTRIBUTING.md#the-exercise-data).  If you're just getting started and looking for an easy way to get involved, take a look at  [regenerating the test suites](#regenerating-a-test-suite), [porting an exercise from another language](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md), or [creating an automated test generator](#implementing-a-generator).
```
Comment (Insti) [inline README.md]: Add your new section as a new paragraph rather than continuing the line.

Comment (Insti) [reply README.md]: I'm not sure these things are easy ways to get involved.
They are things that are helpful, but can get quite complicated.

File: README.md [hunk 7]
```diff
@@ -33,9 +33,7 @@ A few exercises use a custom test template:
 
 ### BookKeeping::VERSION
 
-For some, even perhaps many, of the exercises, you will find a
-reference to the `BookKeeping` module, but this is only included when
-tests have been generated; see [Generated Test Suites](#generated-test-suites).
+For some, even perhaps many, of the exercises, you will find a reference to the `BookKeeping` module, but this is only included when tests have been generated; see [Generated Test Suites](#generated-test-suites).  This `VERSION` number helps make sure exercise solvers and exercise reviewers know which revision of the test suite they are talking about, and it theoretically helps avoid reviewer feedback like *"Your solution doesn't make the tests pass."*.
```
Comment (Insti) [inline README.md]: > helps avoid reviewer feedback like *"Your solution doesn't make the tests pass."*.

If they were looking at a different version of the tests.

**Comment by Insti** [comment]
Thanks @rpalo this looks on-the-whole great.
I've added comments about the things that caught my eye as a I read through.
The long lines without linebreaks made reviewing particular sections tricky, but that's probably GitHub's fault.
Spelling/grammar, use of apostrophes,  and inclusion of relevant information were all good. 👍

**Comment by rpalo** [comment]
Let me know what you think.  I pulled in all of your feedback.  I didn't even think about the line-breaks thing.  I can go back and split things onto shorter lines if you like.  I originally put paragraphs all on the same line because all of the extra line-breaks made my Markdown editor grumpy, but it's not a big deal to change it back.

**Comment by kotp** [comment]
@rpalo What are you using for a "Markdown editor"?  Nothing really to do with this, but just curious in tooling.

**Comment by kotp** [comment]
exercism/problem-specifications#19 for reference.  I know I appreciate the 80 character "general" limit.

**Comment by rpalo** [comment]
I use [typora](https://typora.io/). Especially for writing longer form things like blog posts and tutorials, you can’t beat its ease of use and beautiful interface.

I’m starting to see the benefits of the 80 char limit as I do more with open source — at least for things meant to be read on github or in the terminal. I’ll go through later and fix that for this pull request.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -1,40 +0,0 @@
-<!--- This template is meant to help develop good habits in creating PRs. -->
```
Comment (Insti) [inline .github/PULL_REQUEST_TEMPLATE.md]: This file should not be part of this Pull Request.

Comment (Insti) [reply .github/PULL_REQUEST_TEMPLATE.md]: I think merging from master was the problem. 
You might like to try a `git rebase -i ` and removing the commits you did not make from this branch.

**Comment by rpalo** [comment]
Oy.  I think I fixed it?  From as far as I can tell?  I'm still figuring out how rebasing and keeping branches up to date and things like that work.  But I'm pretty sure that all that's changing now is:
1. Add new sections that talk about regenerating tests.
2. Incorporate all of @Insti's feedback on my changes.
3. Rebreak all lines at ~80 chars.

😬   Thanks for your patience and help, I really appreciate it.

**Comment by Insti** [comment]
Git can sometimes get a bit out of control 😄 
But it looks like you got it fixed @rpalo , thanks.

File: README.md [hunk 9]
```diff
@@ -104,12 +109,67 @@ tree -L 1 ~/code/exercism
...
 
-Leaving out the --update option will cause the BookKeeping version number to remain the same.
-This can be useful when testing generators.
+```bash
+bundle install
+```
+
+Be sure that you're working on the most up-to-date version of the repo.
+From the root of your copy of the repository:
+
```
Comment (kotp) [inline README.md]: This seems to skip the task of cloning it in the first place.

File: README.md [hunk 10]
```diff
@@ -104,12 +109,67 @@ tree -L 1 ~/code/exercism
...
+```bash
+# Add the exercism repo as upstream if you haven't yet:
+git remote add upstream https://github.com/exercism/ruby.git
+
```
Comment (kotp) [inline README.md]: It could be mentioned that this only needs to be done once.

Comment (Insti) [reply README.md]: Isn't this covered by: 
> if you haven't yet

**Comment by Insti** [comment]
I think it *is* helpful to have the example git commands.

I agree that they could be in a separate document, but until the current Readme becomes too big[1] it is OK that they are there.

[1] It may already be too big.

**Comment by rpalo** [comment]
Hi there!  Just checking in, is there anything more waiting on me that I can do?

**Comment by kotp** [comment]
Nothing for this pull request.

Thank you @rpalo for this.
